editor/src/components/editor/store/dispatch-strategies.tsx

```diff
@@ -563,13 +564,67 @@ function handleStrategiesInner(
   storedState: EditorStoreFull,
   result: InnerDispatchResult,
 ): HandleStrategiesResult {
+  const isSaveDomReport = dispatchedActions.some((a) => a.action === 'SAVE_DOM_REPORT')
+
   const makeChangesPermanent = dispatchedActions.some(shouldApplyClearInteractionSessionResult)
   const cancelInteraction =
     dispatchedActions.some(isClearInteractionSession) && !makeChangesPermanent
   const isInteractionAction = dispatchedActions.some(isCreateOrUpdateInteractionSession)
     ? 'interaction-create-or-update'
     : 'non-interaction'
-  if (storedState.unpatchedEditor.canvas.interactionSession == null) {
+
+  if (isSaveDomReport) {
+    if (
+      // XOR
+      (result.unpatchedEditor.canvas.interactionSession == null) !==
+      (storedState.patchedEditor.canvas.interactionSession == null)
+    ) {
+      throw new Error(
+        'Dispatch error: SAVE_DOM_REPORT changed canvas.interactionSession in an illegal way',
+      )
+    }
+
+    if (
+      // XOR
+      (result.unpatchedEditor.canvas.dragState == null) !==
+      (storedState.patchedEditor.canvas.dragState == null)
+    ) {
+      throw new Error('Dispatch error: SAVE_DOM_REPORT changed canvas.dragState in an illegal way')
+    }
+
+    // the goal here is to keep the old patched EditorState, except for the metadata
+    // Unfortunately there's FIVE places we have metadata in EditorState
+    const oldPatchedEditorWithNewMetadata: EditorState = {
+      ...storedState.patchedEditor, // the "old" patched editor from the action dispatch that triggered SAVE_DOM_WALKER
+      jsxMetadata: result.unpatchedEditor.jsxMetadata, // the fresh metadata from SAVE_DOM_REPORT
+      domMetadata: result.unpatchedEditor.domMetadata,
+      spyMetadata: result.unpatchedEditor.spyMetadata,
+      canvas: {
+        ...storedState.patchedEditor.canvas,
+        interactionSession:
+          result.unpatchedEditor.canvas.interactionSession == null ||
```

Comment:
I try to wrap my head around this, but it is not clear for me why do we have to delete the interactionSession from the old patched editor when the new unpatchededitor doesn't have interactionsession.
Shouldn't we just leave it as it is in these cases?

Reply:
that's why I throw the error a few lines above this if the null-ness of the interaction sessions doesn't match